hera_cal/smooth_cal.py

```diff
@@ -0,0 +1,390 @@
+
+    Returns:
+        conv_gains: gains conolved with a Gaussian kernel in time
+    '''
+    
+    padded_gains, padded_wgts = deepcopy(gains), deepcopy(wgts)
+    for n in range(nMirrors+1):
+        nInt, nFreq = padded_gains.shape
+        kernel = time_kernel(nInt, np.median(np.diff(times))*24*60*60, filter_scale=filter_scale)
+        if n < nMirrors:
+            padded_gains = np.vstack((np.flipud(padded_gains),gains,np.flipud(padded_gains)))
+            padded_wgts = np.vstack((np.flipud(padded_wgts),gains,np.flipud(padded_wgts)))
+
+    conv_gains = padded_gains * padded_wgts**2
+    conv_weights = padded_wgts**2
+    for i in range(nFreq):
+        conv_gains[:,i] = scipy.signal.convolve(conv_gains[:,i], kernel, mode='same')
+        conv_weights[:,i] = scipy.signal.convolve(conv_weights[:,i], kernel, mode='same')
+    conv_gains /= conv_weights
+    conv_gains[np.logical_not(np.isfinite(conv_gains))] = 0
+    return conv_gains[nMirrors*len(times):(nMirrors+1)*len(times), :]
+
+
+
+class Calibration_Smoother():
+
+    def __init__(self):
+        '''TODO: EDIT THIS Class for loading data, performing uvtools.dspec.delay_filter, and writing out data using pyuvdata.
+        To use, run either self.load_data() or self.load_dicts(), then self.run_filter(). If data is loaded with a single 
+        string path or a single UVData object, it can be written to a new file using self.write_filtered_data().
+        '''
+        self.has_cal, self.has_prev_cal, self.has_next_cal = False, False, False
+        self.has_data, self.has_prev_data, self.has_next_data = False, False, False
+
+
+    def reset_filtering(self):
+        '''Reset gain smoothing to the original input gains.'''
+        self.filtered_gains = deepcopy(self.gains)
+        self.freq_filtered, self.time_filtered = False, False
+
+
+    def check_consistency(self):
+        '''After loading both the data and the calibration solutions, this function 
+        makes sure everything (times, freqs, ants, etc.) lines up.'''
+        if not (self.has_cal and self.has_data):
+            raise AttributeError('Data consistency cannot be checked unless both load_cal() and load_data() have been run.')
+
+        assert(len(self.times) == len(self.data_times))
+        #assert(self.times == self.data_times) #TODO: revisit this
+        assert(np.all(self.freqs == self.data_freqs))
+        for (ant,pol) in self.gains.keys():
+            assert((ant, ant, pol+pol) in self.data) #assert data has autocorrelations
+        
+        if self.has_prev_cal or self.has_prev_data:
+            assert(self.has_prev_cal and self.has_prev_data) #should have both or it doesn't make sense
+            assert(np.all(self.prev_freqs == self.freqs))
+            assert(np.all(self.prev_data_freqs == self.freqs))
+            assert(len(self.prev_times) == len(self.prev_data_times))
+            #assert(np.all(self.prev_times == self.prev_data_times)) #TODO: revisit this
```

Comment:
lines 185 and (the commented) 186 don't seem to be a necessary check. The way your time filter is written, it can handle any size prev_time and next_times. I would suggest instead inserting a check that the times are contiguous between prev/current/next.

Reply:
It is necessary that the data files correspond to the same times as the calibration files. I can't pass that check (and I opened issue #204), so I'm settling for a length check. While len(self.prev_times) does not need to be the same len(self.times), it does need to be the same as len(self.prev_data_times)

I've added in contiguity checks you requested.